internal/ingress/controller/kong.go

```diff
@@ -112,11 +113,44 @@ func (n *NGINXController) OnUpdate(ingressCfg *ingress.Configuration) error {
 	return nil
 }
 
+// Duplicated by github.com/kubernetes/ingress-nginx/internal/ingress/controller/controller.go#getServiceClusterEndpoint
+func getServiceClusterEndpoint(svcKey string, backend *ingress.Backend) (endpoint ingress.Endpoint, err error) {
+	svc := backend.Service
+
+	if svc.Spec.ClusterIP == "" || svc.Spec.ClusterIP == "None" {
+		return endpoint, errors.Errorf("No ClusterIP found for service %s", svcKey)
+	}
+
+	endpoint.Address = svc.Spec.ClusterIP
+
+	// If the service port in the ingress uses a name, lookup
+	// the actual port in the service spec
+	if backend.Port.Type == intstr.String {
+		var port int32 = -1
+		for _, svcPort := range svc.Spec.Ports {
+			if svcPort.Name == backend.Port.String() {
+				port = svcPort.Port
+				break
+			}
+		}
+		if port == -1 {
+			return endpoint, errors.Errorf("no port mapped for service %s and port name %s", svc.Name, backend.Port.String())
+		}
+		endpoint.Port = fmt.Sprintf("%d", port)
+	} else {
+		endpoint.Port = backend.Port.String()
+	}
+
+	return endpoint, nil
+}
+
 // syncTargets reconciles the state between the ingress controller and
 // kong comparing the endpoints in Kubernetes and the targets in a
 // particular kong upstream. To avoid downtimes we create the new targets
 // first and then remove the killed ones.
-func syncTargets(upstream string, ingressEndpopint *ingress.Backend, client *kong.RestClient) error {
+// If useServiceUpstream set, use service's Cluster IP and port, instead of endpoints.
+// https://github.com/kubernetes/ingress-nginx/blob/master/docs/user-guide/nginx-configuration/annotations.md#service-upstream
+func syncTargets(upstream string, ingressEndpopint *ingress.Backend, client *kong.RestClient, useServiceUpstream bool) error {
```

Comment:
`useServiceUpstream` doesn't convey anything meaningful. Perhaps, a better name would be `useClusterIP`?

Also, there is typo in `ingressEndpopint`. Could you correct that? I understand that the typo was not introduced in this PR. It'll be acceptable if you don't correct it as well. Thanks!

Reply:
Also, could you remove the link to the ingress-nginx reference? Thanks!